esp-hal/src/i2s/master/camera.rs

```diff
@@ -0,0 +1,470 @@
+        self
+    }
+
+    /// Configures the d4 pin.
+    /// Use either [`Self::with_data_pins`] *or*
+    /// [`Self::with_d0`]..[`Self::with_d7`].
+    pub fn with_d4<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::din4_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the d5 pin.
+    /// Use either [`Self::with_data_pins`] *or*
+    /// [`Self::with_d0`]..[`Self::with_d7`].
+    pub fn with_d5<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::din5_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the d6 pin.
+    /// Use either [`Self::with_data_pins`] *or*
+    /// [`Self::with_d0`]..[`Self::with_d7`].
+    pub fn with_d6<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::din6_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the d7 pin.
+    /// Use either [`Self::with_data_pins`] *or*
+    /// [`Self::with_d0`]..[`Self::with_d7`].
+    pub fn with_d7<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::din7_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the pixel clock pin for the camera interface.
+    pub fn with_ws<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::ws_in_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the vertical sync (VSYNC) pin for the camera interface.
+    pub fn with_vsync<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::v_sync_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the horizontal sync (HSYNC) pin for the camera interface.
+    pub fn with_hsync<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::h_sync_signal().connect_to(pin);
+        self
+    }
+
+    /// Configures the horizontal sync enable pin for the camera interface.
+    pub fn with_henable<PIN: InputPin>(self, pin: impl Peripheral<P = PIN>) -> Self {
+        crate::into_mapped_ref!(pin);
+        pin.init_input(Pull::None);
+        I2S::h_enable_signal().connect_to(pin);
+        self
+    }
+}
+
+impl<'d, I2S> Camera<'d, I2S>
+where
+    I2S: RegisterAccess,
+{
+    /// Starts a DMA transfer to receive data from the camera peripheral.
+    pub fn receive<BUF: DmaRxBuffer>(
+        mut self,
+        mut buf: BUF,
+        len: usize,
+    ) -> Result<CameraTransfer<'d, I2S, BUF>, Error> {
```

Comment:
When this fails, along with the error, the camera and the buffer should be returned.

Like `Result<CameraTransfer<'d, I2S, BUF>, (Error, Self, BUF)>`